B mateo prescreening task

BMateo opened the pull request:
**Pull Request Summary**

This PR adds the ability for a staker to delegate their rewards if the latter are supposed to go to the free balance of the staker.
Also a delegated account can delegate to someone else and the staker delegate to himself.

**Check list**
- [x] adds new feature
- [x] modifies existing feature (bug fix or improvements)
- [x] relies on other tasks
- [ ] documentation changes
- [x] tests and/or benchmarks are included

**This pull request makes the following changes:**

**Adds**
- Added storage type DelegatedAccounts
- Added extrinsic to delegate rewards (set_delegated_account)
- Added unit test cheking that this new feature dont break anything and behaves correctly increasing the balance of the delegatee
- Added unit test to verify that only a staker or delegatee can delegate for the staker

**Fixes**
- In claim_staker the free balance of the delegated account its increased if RewardDestination = FreeBalance otherwise the staker free balance its increased.
- In the all function of the MemorySnapshot struct a validation was added to return a MemorySnapshot with the free balance of the delegated or the staker in order to use the assert_restake_reward util that is used in the claim_delegated_account unit test.

**Changes**
- Change mod.rs
- Change testing_utils.rs
- Change test.rs

**To-dos**
- [ ] Update benchmarks
- [ ] Update documentation

BMateo committed:
Added:
- Storage type DelegatedAccounts double map to
- Event DelegatedAccountSet indicating when changes in the delegatee occurs
- Error to indicate that the caller cannot set a delegatee for a staker
- In claim_staker: condition to increase the free balance correctly after restaking
- Extrinsic set_delegated_account to delegate rewards

BMateo committed:
- Condition to return a MemorySnapshot with the free balance of the delegatee
Nedeed by unit tests in order to assert_restake_reward works with the changes

BMateo committed:
- Unit test to claim with a delegatee setted
- Unit test to verify the accout that set a delegatee

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -795,9 +812,13 @@ pub mod pallet {
                     staker_reward,
                 ));
+
+                T::Currency::resolve_creating(&staker, reward_imbalance);
+            } else {
+                let delegated_account = DelegatedAccounts::<T>::get(&staker, &contract_id).unwrap_or(staker.clone());
+                T::Currency::resolve_creating(&delegated_account, reward_imbalance);
```
Quite elegant.

Does something else needs to be changed/updated related to this call?

BMateo replied:
The Reward event emitted can be changed depending on what account we want to emit in the event, the staker account in wich the rewards has been claimed or the account where the rewards were deposited.

Dinonard replied:
Some tools, like Subscan rely on the Reward event to understand how much some staker has earned. Changing the event like this can have wider consequences. Personally I wouldn't do it but it's a good answer to the question.

What about weight required to execute this call?

BMateo replied:
The worst case scenario for this function is when a user does not restake the rewards and has a delegated setted. In that case the function needs to read the storage one aditional time.

Dinonard commented on frame/dapps-staking/src/pallet/mod.rs:
```diff
@@ -978,6 +999,35 @@ pub mod pallet {
+
+            Self::deposit_event(Event::<T>::DelegatedAccountSet(staker, contract_id, delegated_account));
+            Ok(().into())
+        } 
```
What if the original staker wants to remove a delegate? I.e. if staker wants to receive rewards into their own account.

BMateo replied:
The staker should set as the delegatee himself but an extrinsic can be added to clean the storage map for the staker

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -15,19 +15,32 @@ pub(crate) struct MemorySnapshot {
+                staker_info: GeneralStakerInfo::<TestRuntime>::get(&account, contract_id),
+                contract_info: DappsStaking::contract_stake_info(contract_id, era).unwrap_or_default(),
+                ledger: DappsStaking::ledger(&account),
+                free_balance: <TestRuntime as Config>::Currency::free_balance(&delegated_account),
```
This is hacky, and you loose the opportunity to verify that free balance of staker hasn't been changed in case of delegated reward.

Also, lots of code duplication that can be avoided.

BMateo replied:
I agree on that missing case. I left the MemorySnapshot as it was but changed the assert_restake_reward function in order to assert correctly when the rewards were delegated.

Dinonard replied:
Why not add `delegate` and `delegate_free_balance` to the struct?

How would you add it to easily convey the information in case delegate exists, and in case delegate doesn't exist? What type would you use?

BMateo replied:
I was thinking on doing it that way but I choose to add parameters to assert_restake_reward function.
The other alternative was to add the field to the struct using an Option type and then asserting if a value is available.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -2172,3 +2172,137 @@ pub fn set_contract_stake_info() {
+        assert_bond_and_stake(first_staker, &second_contract_id, 33);
+        assert_bond_and_stake(second_staker, &second_contract_id, 22);
+    
+        let eras_advanced = 3;
+        advance_to_era(start_era + eras_advanced);
+    
+        for x in 0..eras_advanced.into() {
+            assert_bond_and_stake(first_staker, &first_contract_id, 20 + x * 3);
+            assert_bond_and_stake(second_staker, &first_contract_id, 5 + x * 5);
+            advance_to_era(DappsStaking::current_era() + 1);
+        }
```
Can you explain why such scenario is needed (tech.requirement or a personal preference)?

BMateo replied:
Its a personal preference. I dont think the scenario is needed but I wanted to make sure that the changes to the code dont break the existing features and test cases.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -2172,3 +2172,137 @@ pub fn set_contract_stake_info() {
+            assert_claim_dapp(&first_contract_id, era);
+            assert_claim_staker(second_staker, &first_contract_id);
+        }
+    
+        // Shouldn't be possible to claim current era.
+        // Also, previous claim calls should have claimed everything prior to current era.
+        assert_noop!(
+            DappsStaking::claim_staker(Origin::signed(first_staker), first_contract_id.clone()),
+            Error::<TestRuntime>::EraOutOfBounds
+        );
+        assert_noop!(
+            DappsStaking::claim_dapp(
+                Origin::signed(first_developer),
+                first_contract_id,
+                current_era
+            ),
+            Error::<TestRuntime>::EraOutOfBounds
+        );
+    })
```
Is this relevant for this feature?

BMateo replied:
It is not relevant to claim dapp right now. I updated the test removing unnecessary lines and adding assertions to the delegatee free balance.

Dinonard commented on frame/dapps-staking/src/tests.rs:
```diff
@@ -2172,3 +2172,137 @@ pub fn set_contract_stake_info() {
+        assert_register(developer, &contract_id);
+        assert_bond_and_stake(staker, &contract_id, 100);
+    
+        let eras_advanced = 3;
+        advance_to_era(start_era + eras_advanced);
```
Why is this required for this unit test?

BMateo replied:
It is not required

Dinonard reviewed (commented):
Nice and simple solution.

Left you some comments and I'd like to know your opinion & comments.

Also, if this was going into production (on live chain), what else would need to be done?

BMateo committed:
-Remove unnecesary instructions

BMateo committed:
Added:
- DelegatedAccountRemoved event
- CannotDelegateToStaker and NoDelegateeSet errors
- Modification in set_delegated_account. So a staker cannot be setted as his own delegatee
- Extrinsic to remove the delegatee for a staker

BMateo committed:
- test cases for the changes in remove_delegate and set_delegated_account

BMateo committed:
- restore the all function from MemorySnapshot
- Modifications to function assert_restake_reward to assert the free balance of the staker when the rewards are delegated

BMateo committed:
- remove unnecesary instructions and assert that the free balance of the delegatee increases after claiming

BMateo commented:
> Also, if this was going into production (on live chain), what else would need to be done?
If the functionality, testing, benchmarking and docs are okay one thing that Im missing its to update the version of the pallet.

BMateo committed:
Update pallet version

BMateo committed:
Update precompiles dapps-staking version

BMateo committed:
- Change in the parameter of Reward event emitted

BMateo committed:
- Assert the delagated account emited for the Reward event if rewards are delegated

Dinonard commented on frame/dapps-staking/src/testing_utils.rs:
```diff
@@ -538,6 +526,8 @@ pub(crate) fn assert_claim_staker(claimer: AccountId, contract_id: &MockSmartCon
         &init_state_current_era,
         &final_state_current_era,
         calculated_reward,
+        claimer.clone(),
+        contract_id.clone(),
```
That's a lot of args, for a function that was already maybe too comlex :slightly_smiling_face:

BMateo replied:
Got it, I will look for an alternative.

BMateo committed:
- Remove event modifications, the staker its always emited in Reward event

BMateo committed:
- Added fields to MemorySnapshot
- Removed extra parameters to assert_restake_reward
- Added assertion to the free balance of delegated
